File: Leetcode_JS/problem1979.js
function getHCF(dividend, divisor) {
    while (divisor != 0) {
        const rem = dividend % divisor;
        dividend = divisor;
        divisor = rem;
    }

    return dividend;
}

// optimal
function findGCD(nums) {
    let largest = nums[0];
    let smallest = nums[0];

    for (const n of nums) {
        if (n > largest)
            largest = n;
        else if (n < smallest)
            smallest = n;
    }

    return getHCF(largest, smallest);
}


// Test Cases
// console.log(findGCD([1, 1]));
// console.log(findGCD([1, 2]));
// console.log(findGCD([2, 5, 6, 9, 10]));
// console.log(findGCD([4, 5, 6, 8, 12]));
// console.log(findGCD([2, 5, 6, 9, 10]));
// console.log(findGCD([7, 11, 13]));
// console.log(findGCD([8, 15, 22]));
// console.log(findGCD([42, 42, 42]));

File: Leetcode_CPP/problem1979.cpp
#include <iostream>
#include <stdio.h>
#include <vector>
using namespace std;

int getGCD(int a, int b) {
    int divident = a > b? a: b;
    int divisor = a < b? a: b;

    while (divisor != 0)
    {
        int rem = divident % divisor;
        divident = divisor;
        divisor = rem;
    }
    
    return divident;
}

int findGCD(vector<int> &nums)
{
    int smallest = 1000;
    int biggest = 0;

    for (int i = 0; i < nums.size(); i++)
    {
        if (nums[i] < smallest)
            smallest = nums[i];
        
        if (nums[i] > biggest)
            biggest = nums[i];
    }

    return getGCD(smallest, biggest);
}

int main()
{
    vector<int> nums = {10, 6, 9};
    findGCD(nums);
    return 0;
}
